core/block_dragger.js

```diff
@@ -256,46 +279,73 @@ Blockly.BlockDragger.prototype.endBlockDrag = function(e, currentDragDeltaXY) {
     this.draggingBlock_.dispose(false, true);
     Blockly.draggingConnections = [];
   } else {
-    // These are expensive and don't need to be done if we're deleting.
-    this.draggingBlock_.moveConnections(delta.x, delta.y);
-    this.draggingBlock_.setDragging(false);
-    this.fireMoveEvent_();
-    if (this.draggedConnectionManager_.wouldConnectBlock()) {
-      // Applying connections also rerenders the relevant blocks.
-      this.draggedConnectionManager_.applyConnections();
-    } else {
-      this.draggingBlock_.render();
-    }
-    this.draggingBlock_.scheduleSnapAndBump();
+    // Moving the block is expensive, so only do it if the block is not deleted.
+    this.updateBlockAfterMove_(delta);
   }
   this.workspace_.setResizesEnabled(true);
 
-  var toolbox = this.workspace_.getToolbox();
-  if (toolbox && typeof toolbox.removeStyle == 'function') {
-    var style = this.draggingBlock_.isDeletable() ? 'blocklyToolboxDelete' :
-        'blocklyToolboxGrab';
-    toolbox.removeStyle(style);
+  this.updateToolboxStyle_(true);
+};
+
+/**
+ * Updates the necessary information to place a block at a certain location.
+ * @param {!Blockly.utils.Coordinate} delta The change in location from where
+ *     the block started the drag to where it ended the drag.
+ * @protected
+ */
+Blockly.BlockDragger.prototype.updateBlockAfterMove_ = function(delta) {
+  this.draggingBlock_.moveConnections(delta.x, delta.y);
+  this.draggingBlock_.setDragging(false);
+  this.fireMoveEvent_();
+  if (this.draggedConnectionManager_.wouldConnectBlock()) {
+    // Applying connections also rerenders the relevant blocks.
+    this.draggedConnectionManager_.applyConnections();
+  } else {
+    this.draggingBlock_.render();
   }
-  Blockly.Events.setGroup(false);
+  this.draggingBlock_.scheduleSnapAndBump();
 };
 
 /**
  * Fire a UI event at the end of a block drag.
- * @private
+ * @protected
  */
 Blockly.BlockDragger.prototype.fireDragEndEvent_ = function() {
   var event = new (Blockly.Events.get(Blockly.Events.BLOCK_DRAG))(
       this.draggingBlock_, false, this.draggingBlock_.getDescendants(false));
   Blockly.Events.fire(event);
 };
 
+/**
+ * Adds or removes the style of the cursor for the toolbox.
+ * This is what changes the cursor to display an x when a deletable block is
+ * held over the toolbox.
+ * @param {boolean} isEnd True if we are at the end of a drag, false otherwise.
+ * @protected
+ */
+Blockly.BlockDragger.prototype.updateToolboxStyle_ = function(isEnd) {
+  var toolbox = this.workspace_.getToolbox();
+
+  if (toolbox) {
```

Comment:
I know this was already here and you just moved it into a function. So we don't have to change anything rn. But it seems weird that the BlockDragger does this. It should just notify the toolbox and let the toolbox worry about its own styles.

Once we have delete area components, would this change to accommodate all delete areas? Then it makes even more sense, for each delete area, call some method on it, and let them figure out what to do. I'm not actually sure if this does matter for all delete areas though. There is the somewhat similar `updateCursorDuringBlockDrag` and I'm not sure how they are related/different.

Reply:
Do you need this method to be `protected` (do you override it)? I think it could be potentially refactored into the toolbox using the `IDragTarget` API.

Reply:
I was able to get this logic factored out into the `Toolbox` class using the changes in https://github.com/google/blockly/pull/4888. I can make a follow up PR after this one so we can discuss pulling it out.